### PR TITLE
decoder: add flush() for h264 and h265

### PR DESCRIPTION
**Pull request by LindaYu17**
add flush() function for h264 and h265, so that we could flush dpb for
them in this function.

Signed-off-by: Linda Yu linda.yu@intel.com

### CODE REVIEW

File: decoder/vaapidecoder_h264.cpp [hunk 1]
```diff
@@ -1783,6 +1783,12 @@ YamiStatus VaapiDecoderH264::decodeNalu(NalUnit* nalu)
     return status;
 }
 
+void VaapiDecoderH264::flush(void)
+{
+    m_dpb.flush();
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h264.cpp:1825]: could you move all things between https://github.com/01org/libyami/blob/apache/decoder/vaapidecoder_h264.cpp#L1825 and  https://github.com/01org/libyami/blob/apache/decoder/vaapidecoder_h264.cpp#L1830 to flush? and call flush in L1825

File: decoder/vaapidecoder_h265.cpp [hunk 2]
```diff
@@ -1068,6 +1068,14 @@ YamiStatus VaapiDecoderH265::decodeNalu(NalUnit* nalu)
     return status;
 }
 
+void VaapiDecoderH2665::flush(void)
+{
+    m_dpb.flush();
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp:1088]: same as h264.

File: decoder/vaapidecoder_h265.cpp [hunk 3]
```diff
@@ -1068,6 +1068,14 @@ YamiStatus VaapiDecoderH265::decodeNalu(NalUnit* nalu)
     return status;
 }
 
+void VaapiDecoderH2665::flush(void)
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: this not build able

File: decoder/vaapidecoder_h264.cpp (hunk 1 above)
Comment (xuguangxin) [reply decoder/vaapidecoder_h264.cpp:1825]: Do we need reset parser here?

**Review by xuguangxin — CHANGES_REQUESTED** [review]

File: decoder/vaapidecoder_h264.cpp (hunk 1 above)
Comment (xuguangxin) [reply decoder/vaapidecoder_h264.cpp:1825]: Please check  following things need reset or not.
    YamiParser::H264::Parser m_parser;
    PicturePtr m_currPic;
    PicturePtr m_prevPic;
    bool m_newStream;
    bool m_endOfSequence;
    bool m_endOfStream;
    DPB m_dpb;
    uint32_t m_nalLengthSize;
    SurfacePtr m_currSurface;
    bool m_contextChanged;
    static const bool s_registered; // VaapiDecoderFactory registration resu

File: decoder/vaapidecoder_h265.cpp (hunk 2 above)
Comment (xuguangxin) [reply decoder/vaapidecoder_h265.cpp:1088]: Please check following fields too
    SharedPtr<Parser> m_parser;
    PicturePtr  m_current;
    uint16_t    m_prevPicOrderCntMsb;
    int32_t     m_prevPicOrderCntLsb;
    int32_t     m_nalLengthSize;
    bool        m_associatedIrapNoRaslOutputFlag;
    bool        m_noRaslOutputFlag;
    bool        m_newStream;
    bool        m_endOfSequence;
    DPB         m_dpb;
    std::map<int32_t, uint8_t> m_pocToIndex;
    SharedPtr<SliceHeader> m_prevSlice;

File: decoder/vaapidecoder_h265.cpp [hunk 4]
```diff
@@ -1083,15 +1082,26 @@ YamiStatus VaapiDecoderH265::decodeNalu(NalUnit* nalu)
...
 
+void VaapiDecoderH265::flush(void)
+{
+    decodeCurrent();
+    m_dpb.flush();
+    m_prevPicOrderCntMsb = 0;
+    m_prevPicOrderCntLsb = 0;
+    m_newStream = true;
+    m_endOfSequence = false;
+    m_associatedIrapNoRaslOutputFlag = false;
```
Comment (xuguangxin) [inline decoder/vaapidecoder_h265.cpp]: we do not need change this:

```
+    m_associatedIrapNoRaslOutputFlag = false;
 +    m_noRaslOutputFlag = false;
 +    m_current.reset();
 +    m_pocToIndex.clear();
```

you need change this  
`+    m_prevSlice.reset();`
to
`m_prevSlice.reset(new SliceHeader());`

Comment (xuguangxin) [reply decoder/vaapidecoder_h265.cpp]: @lizhong1008 , Please help to review this pull request.
thanks

**Review by xuguangxin — CHANGES_REQUESTED** [review]

File: decoder/vaapidecoder_h264.cpp [hunk 5]
```diff
@@ -1819,14 +1819,24 @@ YamiStatus VaapiDecoderH264::decodeNalu(NalUnit* nalu)
...
+void VaapiDecoderH264::flush(void)
+{
+    decodeCurrent();
+    m_dpb.flush();
+    m_newStream = true;
+    m_endOfStream = false;
+    m_endOfSequence = false;
+    m_currPic.reset();
+    m_prevPic.reset();
+    m_currSurface.reset();
```
Comment (lizhong1008) [inline decoder/vaapidecoder_h264.cpp:1831]: @xuguangxin , do we need implement the stop() interface, and move the sw/hw context resetting to stop()?

Comment (xuguangxin) [reply decoder/vaapidecoder_h264.cpp:1831]: But flush is not stop, it do not need deinit Va. Maybe we can put all context reset thing to a standalone function like resetDecoderContext.

Comment (lizhong1008) [reply decoder/vaapidecoder_h264.cpp:1831]: I mean we don't need reset surface in flush(), maybe we can move all surface and context into stop().

Comment (xuguangxin) [reply decoder/vaapidecoder_h264.cpp:1831]: flush is for seek, if we do not reset m_currSurface, we will reference error surface after seek.

**Comment by xuguangxin** [comment]
Hi @lizhong1008 , is it ok to start verification this?

**Review by xuguangxin — APPROVED** [review]

**Comment by xuguangxin** [comment]
Hi @lizhong1008 , is it ok to start verification this?

**Comment by lizhong1008** [comment]
@xuguangxin LGTM

**Comment by xuguangxin** [comment]
@wudping , Please help to verify this on h264 and h265 conformance.

**Comment by wudping** [comment]
I am sorry, There is a problem in 264/265 decoding with this patch. Some frames are lost.
My test file is alphaconformanceG.264.
host: regression@regression,
pw: 123456
path: /home/regression/daily_regression/medias_for_verify/mediafiles/h264/alphaconformanceG.264

**Comment by xuguangxin** [comment]
@wudping do not paste you password in github ;), you can sent the clip to @LindaYu17 using email.
thanks

**Comment by wudping** [comment]
okay, thanks @xuguangxin .

**Comment by wudping** [comment]
I am sorry, there is a problem in decoding 265/264 with this patch. Some frames are lost.

**Comment by xuguangxin** [comment]
Hi @LindaYu17 : 
Do you have time estimate for this pull request?

**Comment by LindaYu17** [comment]
will make time early next week on this!

**Comment by LindaYu17** [comment]
hi @xuguangxin I have updated this pull request.
hi @wudping could you please help to verify again?

**Comment by wudping** [comment]
okay, @LindaYu17 .

**Comment by wudping** [comment]
verify + 1

**Comment by xuguangxin** [comment]
merged as 9eba4cfb4f9c3eb7757c421c5c37a723d37a0625. thanks Linda

**Pull request closed without merging**